tests/objectIndex.cc

```diff
@@ -0,0 +1,179 @@
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the LSST License Statement and
+ * the GNU General Public License along with this program.  If not,
+ * see <http://www.lsstcorp.org/LegalNotices/>.
+ */
+#define BOOST_TEST_DYN_LINK
+#define BOOST_TEST_MODULE ObjectIndex
+#include "boost/test/unit_test.hpp"
+
+#include "lsst/partition/Chunker.h"
+#include "lsst/partition/Csv.h"
+#include "lsst/partition/ObjectIndex.h"
+#include "TempFile.h"
+
+#include <stdexcept>
+#include <string>
+#include <vector>
+
+using namespace lsst::partition;
+
+BOOST_AUTO_TEST_CASE(ObjectIndexTest) {
+
+    BOOST_REQUIRE_NO_THROW({
+        ObjectIndex::instance();
+    });
+
+    // It should be safe to close an index when it's not open
+    BOOST_REQUIRE_NO_THROW({
+        ObjectIndex::instance().close();
+    });
+    BOOST_REQUIRE_NO_THROW({
+        ObjectIndex::instance().isOpen();
+    });
+    BOOST_CHECK_EQUAL(ObjectIndex::instance().isOpen(), false);
+    BOOST_REQUIRE_NO_THROW({
+        ObjectIndex::instance().mode();
+    });
+
+    // Test for a failure to use an index while it's not open
+    BOOST_CHECK_THROW({
+        std::string const id = "12345";
+        ObjectIndex::instance().read(id);
+    }, std::logic_error);
+
+    // Test for a failure to use an index if it was not created
+    BOOST_CHECK_THROW({
+        std::string const id = "12345";
+        int32_t const chunkId = 1;
+        int32_t const subChunkId = 2;
+        bool const overlap = false;
+        ChunkLocation const location(chunkId, subChunkId,overlap);
+        ObjectIndex::instance().write(id, location);
+    }, std::logic_error);
+
+    TempFile t;
+    std::string const indexFileName = t.path().string();
+    std::string const indexUrl = "file:" + indexFileName;
+
+    std::string const null = "\\N";
+    char const delimiter = ',';
+    char const escape    = '\\';
+    char const noQuote   = '\0';
+    csv::Dialect const dialect(null, delimiter, escape, noQuote);
+    std::vector<std::string> const fields = {"id", "chunkId", "subChunkId"};
+    csv::Editor editor(dialect, dialect, fields, fields);
+
+    // Make sure input parameters are properly valudated
+    BOOST_CHECK_THROW({
+        ObjectIndex::instance().create("",            editor, "id", "chunkId", "subChunkId");
+        ObjectIndex::instance().create(indexFileName, editor, "",   "chunkId", "subChunkId");
+        ObjectIndex::instance().create(indexFileName, editor, "id", "",        "subChunkId");
+        ObjectIndex::instance().create(indexFileName, editor, "id", "chunkId", "");
+        ObjectIndex::instance().open("",      dialect);
+        ObjectIndex::instance().open("file:", dialect);
```

Comment:
Which one these should throw? If each of them than you need separate test for each.

Reply:
Fixed